Extend 'flutter symbolize' to handle deferred loading units.

sstrickl opened the pull request:
Adds `-u`/`--unit-id-debug-info` arguments to `flutter symbolize` to pass paths to DWARF information for deferred loading units. The argument passed via `-u` should be of the form `N:P`, where `N` is the loading unit ID (an integer) and `P` is the path to the debug information for loading unit `N`. The DWARF information for the root loading unit can either be passed by `-d`/`--debug-info` as before or by `--unit-id-debug-info 1:<path>`.

Partial fix for https://github.com/flutter/flutter/issues/137527. Additional work is needed to adjust tools built on top of `flutter symbolize` to store and pass along this additional information appropriately when there are deferred loading units.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

sstrickl committed:
Extend 'flutter symbolize' to handle deferred loading units.

sstrickl committed:
Add root loading unit id to error message in this case.

sstrickl committed:
Better error messages, test these cases.

christopherfujino commented on packages/flutter_tools/lib/src/commands/symbolize.dart:
```diff
@@ -63,16 +72,50 @@ class SymbolizeCommand extends FlutterCommand {
+        throwToolExit('Different paths were given for the same loading unit'
+            ' $unitId: "${map[unitId]}" and "$unitDebugPath".');
+      }
+      map[unitId] = unitDebugPath;
```
can we/should we validate this is a valid path? Not sure how to do that, but if it's just gonna fail later on, it might be helpful to fail here.

sstrickl replied:
Done, moved the checking on the map paths that was originally only in `validateCommand` and not in `runCommand` to this method instead. (Also have it returning a map of `File` references instead of `String`s now.

christopherfujino commented on packages/flutter_tools/lib/src/commands/symbolize.dart:
```diff
@@ -190,13 +264,37 @@ class DwarfSymbolizationService {
+        input: input,
+        output: output,
+        unitSymbols: <int, Uint8List>{
+          1: symbols,
```
why hardcode the 1 here?

sstrickl replied:
So for an application with deferred units, the root loading unit (e.g., the one that contains the starting point of the application) always has ID 1.

`rootLoadingUnitId` is defined in `lib/src/constants.dart`, but I don't explicitly export any part of that as part of `pkg:native_stack_traces`. While we could just import that file directly to get to it, I'll make an update that officially exports it for uses outside of the Dart repo instead.

sstrickl replied:
Oh, I see, I made a constant for this locally but didn't use it here. In any case, I think it should get that constant from `pkg:native_stack_traces` anyway. (I don't expect it to ever change, but you know...)

sstrickl replied:
Okay, constant used here as well.

christopherfujino commented on packages/flutter_tools/lib/src/commands/symbolize.dart:
```diff
@@ -108,47 +151,71 @@ class SymbolizeCommand extends FlutterCommand {
+
+      // If it's a dSYM container, expand the path to the actual DWARF.
+      if (debugInfoPath.endsWith('.dSYM')) {
+        final Directory debugInfoDir = _fileSystem
```
Call me paranoid, but can we check that `debugInfoDir` exists, and then throw some kind of error if it does not? Maybe a `StateError`? I've learned not to assume anything about any file system operation on users' machines.

sstrickl replied:
There's a `MachO.handleDSYM' in `pkg:native_stack_traces` (also not exported) that could be used here instead of manually recreating the same checks. I'll also export that.

sstrickl replied:
Actually, it uses `dart:io` directly instead of going through the `FileSystem` abstraction, so changing my mind on doing this.

sstrickl replied:
However, that leaves the question as to how to appropriately check this using the `package:file` interface, but I imagine that `childDirectory` and `listSync` do enough checking and throwing if invalid that we're covered for this case? (We already check the path is a directory before turning it into a `Directory` and then calling `childDirectory` on it.)

christopherfujino replied:
SGTM

christopherfujino commented on packages/flutter_tools/lib/src/commands/symbolize.dart:
```diff
@@ -108,47 +151,71 @@ class SymbolizeCommand extends FlutterCommand {
+        if (dwarfFiles.length == 1) {
+          debugInfoPath = dwarfFiles.first.path;
+        } else {
+          throwToolExit('Expected a single DWARF file in a dSYM container.');
```
This is maybe a philosophical question, but I'd prefer this not be a ToolExit. `ToolExit` means user error that only they can fix. It's not clear to me if they hit this point, that it is necessarily their fault, nor that they would know what to do (if I hit this, I'd be like, "what does that mean?").

Also ToolExits don't show up in our crash logging, whereas if lots of users hit this we would probably want to know about this. What do you think?

sstrickl replied:
Yeah, you're right, that wouldn't be ideal in either situation then. What would be the best approach for that case? Is there something better for revealing internal errors (e.g., something that's our fault) or should it just `throw`?

sstrickl replied:
Have just used `.single` so that it causes an appropriate `StateError` if it turns out the directory isn't populated by a single file (either empty or more than one).

andrewkolos commented:
@sstrickl Friendly ping.

sstrickl commented:
Thanks for the ping! Sorry, was on leave for a month and then forgot to re-check this once I returned. Looking at Chris's responses now and will address them either code-wise or with words :)

sstrickl committed:
Address comments.

sstrickl commented:
Okay, Christopher, I've taken a shot at addressing all your comments. PTAL and let me know if there's anything remaining, or if the changes don't address them adequately!

christopherfujino reviewed (approved):
LGTM

christopherfujino commented:
@andrewkolos friendly ping

andrewkolos reviewed (approved):
RSLGTM

sstrickl committed:
Clean up tests, get failing tests working.

sstrickl commented:
So looking at the failures, they all are the `.dSYM` tests, which are just checking that if the `.dSYM` package directory exists, are failing. However, I'm not sure why, as we never `createSync` the directory it's looking for in the fresh `MemoryFileSystem` created using `MemoryFileSystem.test()`, but `_fileSystem.directory(fileName).exists[Sync]` and `_fileSystem.isDirectory[Sync]` both return true for it despite it not being there, and so validation succeeds and its not until the run actually tries to read the bytes out of the non-existent file deep within the non-existent .dSYM directory that fails. Hmm.

sstrickl committed:
Fix lack of _handleDSYM on --debug-info argument.

sstrickl committed:
Merge branch 'master' into deferred-symbolize

sstrickl commented:
Coming back to this after a couple weeks OOO and just realized that I never called `_handleDSYM` on the path passed as `--debug-info`, if any. That's why those tests were failing. 🤦‍♀️

sstrickl committed:
Fix analyzer issue.
